### PR TITLE
6: Integrate Vision Into Odometry

### LINKED ISSUE #4: Integrate Vision into Odometry
Find a way to properly localize the robot using CTRE's odometry class and our new cameras.

### PR DESCRIPTION
**Pull request by cjahner1**

### CODE REVIEW

File: src/main/java/frc/robot/constants/drivetrain/DrivetrainConstants.java [hunk 1]
```diff
@@ -8,4 +17,17 @@ public class DrivetrainConstants
     /** Radians per second */
     public static final double MAX_ANGULAR_RATE = Math.PI;
 
+        public static final AprilTagFieldLayout TAG_LAYOUT;
+        
+        static {
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/constants/drivetrain/DrivetrainConstants.java]: fix brackets here

File: src/main/java/frc/robot/constants/drivetrain/DrivetrainConstants.java [hunk 2]
```diff
@@ -8,4 +17,17 @@ public class DrivetrainConstants
...
+            try {
+                TAG_LAYOUT = AprilTagFields.kDefaultField.loadAprilTagLayoutField();
+            } catch (UncheckedIOException e) {
+                e.printStackTrace();
+                throw new RuntimeException(e);
+            }
+        }
+
+        public static final double AMBIGUITY_THRESHOLD = 0.25;
+        public static final Transform3d ROBOT_TO_CAM = new Transform3d(new Translation3d(Units.inchesToMeters(6.676595), 0.0, Units.inchesToMeters(27.63304)), new Rotation3d(0, -45, 0));
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/constants/drivetrain/DrivetrainConstants.java]: create constants for both cameras even if we aren't using it yet. Create unique names (ie. FORWARD_CAMERA_POSITION or REAR_CAMERA_POSITION)

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 3]
```diff
@@ -35,6 +54,13 @@ public Drivetrain(SwerveDrivetrainConstants driveTrainConstants, double Odometry
   public Drivetrain(SwerveDrivetrainConstants driveTrainConstants, SwerveModuleConstants... modules)
   {
     super(driveTrainConstants, modules);
+    // Create a photon camera and pose estimator object
+    m_photonCamera = new PhotonCamera("photonvision");
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: I want to move away from default naming for both cameras choose names that make sense and implement now.

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 4]
```diff
@@ -51,6 +77,43 @@ public Command applyRequest(Supplier<SwerveRequest> requestSupplier)
   @Override
   public void periodic() 
   {
-    // Intentionally Empty
+    //Publish pose for advantagescope odometry
+    m_posePublisher.set(m_odometry.getEstimatedPosition());
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: remove this

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 5]
```diff
@@ -51,6 +77,43 @@ public Command applyRequest(Supplier<SwerveRequest> requestSupplier)
...
+        // Seed pose
+        this.m_odometry.resetPosition(estPose.get().estimatedPose.getRotation().toRotation2d(), m_modulePositions,
+            estPose.get().estimatedPose.toPose2d());
+        m_odometrySeeded = true;
+      } else
+      {
+        // Add vision to kalman filter
+        this.addVisionMeasurement(estPose.get().estimatedPose.toPose2d(), estPose.get().timestampSeconds);
+        SmartDashboard.putString("Pose - Vision", estPose.get().estimatedPose.toPose2d().toString());
+
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: what's up with this white space

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 6]
```diff
@@ -51,6 +77,43 @@ public Command applyRequest(Supplier<SwerveRequest> requestSupplier)
...
+
+      }
+    }
+    // Report robots current pose to smartdashboard every half second
+    if (m_timer.get() > 0.5)
+    {
+      m_timer.reset();
+      SmartDashboard.putString("Pose - Drivetrain", m_odometry.getEstimatedPosition().toString());
+      try {
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: fix your  brackets

**Review by jasonsoohoo — CHANGES_REQUESTED** [review]
looks good needs several adjustments

File: src/main/java/frc/robot/Robot.java [hunk 7]
```diff
@@ -4,11 +4,15 @@
...
+import org.littletonrobotics.junction.LogDataReceiver;
+import org.littletonrobotics.junction.LoggedRobot;
+import org.littletonrobotics.junction.Logger;
+import org.littletonrobotics.junction.networktables.NT4Publisher;
+
 import edu.wpi.first.wpilibj2.command.Command;
 import edu.wpi.first.wpilibj2.command.CommandScheduler;
 
-public class Robot extends TimedRobot {
+public class Robot extends LoggedRobot {
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/Robot.java]: fix the brackets in this file

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 8]
```diff
@@ -51,6 +76,44 @@ public Command applyRequest(Supplier<SwerveRequest> requestSupplier)
...
+    {
+      m_timer.reset();
+      SmartDashboard.putString("Pose - Drivetrain", m_odometry.getEstimatedPosition().toString());
+      try 
+      {
+        SmartDashboard.putNumber("Pose Ambuguity", m_photonCamera.getLatestResult().getBestTarget().getPoseAmbiguity());
+      } 
+      catch(Exception e) 
+      {
+
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: intentionally empty

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 9]
```diff
@@ -51,6 +76,44 @@ public Command applyRequest(Supplier<SwerveRequest> requestSupplier)
...
+      }
+      SmartDashboard.putBoolean("Od seeded", m_odometrySeeded);
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: spell out odometry

**Review by jasonsoohoo — CHANGES_REQUESTED** [review]
some more changes

**Review by jasonsoohoo — APPROVED** [review]

**Review by IronBenderII — APPROVED** [review]
Mid update, but good enough for day 1 of week 2

**Pull request merged**